### PR TITLE
Add metrics for projects using OIDC publishers

### PR DESCRIPTION
**Pull request by sethmlarson**
Track the uptake of OIDC publishers on PyPI.

### CODE REVIEW

File: tests/unit/oidc/test_tasks.py
Comment (miketheman) [inline tests/unit/oidc/test_tasks.py:9]: **question:** ‏Looks like `monkeypatch` fixture isn't used in the test, should it be removed?
```suggestion
def test_compute_oidc_metrics(db_request):
```

File: tests/unit/oidc/test_tasks.py
Comment (miketheman) [inline tests/unit/oidc/test_tasks.py:34]: **suggestion:** ‏we have a general-purpose `metrics` fixture which you can use instead of creating one. 
https://github.com/pypi/warehouse/blob/2241171cf130b853dd534653c0d932dbf5a0db6e/tests/conftest.py#L71C1-L82

File: tests/unit/oidc/test_tasks.py
Comment (miketheman) [inline tests/unit/oidc/test_tasks.py:39]: Using using the fixture, this ought to become:
```suggestion
    assert metrics.gauge.calls == [
```

File: warehouse/oidc/tasks.py
Comment (miketheman) [inline warehouse/oidc/tasks.py:11]: **question:** ‏The query as is is likely to perform sequential scans on the `projects` table which is now in the 400k+ range - we may need to add an index. Have you looked into that already?

**Review by miketheman — COMMENTED** [review]
Looks generally okay, a couple of suggestions and questions inline.

File: warehouse/oidc/tasks.py
Comment (miketheman) [inline warehouse/oidc/tasks.py:17]: Also noticed the usage of `.count()` - which is slower than postgres' `count` function (I know, it's confusing...)

Since we don't intend to use the returned objects in any way, we can use `func.count()` to add the `COUNT(*)` to the SQL query. See https://docs.sqlalchemy.org/en/20/core/functions.html#sqlalchemy.sql.functions.count

**Comment by sethmlarson** [comment]
Thanks for the review @miketheman! I took my inspiration from the [2FA metrics task](https://github.com/pypi/warehouse/commit/0c2ff6d74aa97e31a3d40365fe84f1e491f82dea), I'll get those comments resolved that you mentioned.

File: warehouse/oidc/tasks.py
Comment (miketheman) [reply warehouse/oidc/tasks.py:17]: This is not a hard requirement - we use count() elsewhere already with no material negative impact observed thus far. I saw something, so I said something. 😉

File: warehouse/oidc/tasks.py
Comment (di) [reply warehouse/oidc/tasks.py:11]: Also, technically, this is just counting projects that have a publisher configured, but that doesn't mean they're actually _using_ it.

We might want to count `FileEvents` from OIDC publishers instead:

```sql
select count(*) from file_events where additional->>'publisher_url' is not null;
```

and if we want to count # of projects, we can do the necessary joins from that.

Comment (sethmlarson) [reply warehouse/oidc/tasks.py]: Per @di's suggestion I've distinguished between projects that have configured OIDC and have successfully published a release with OIDC. Hopefully the join-chain is what you were expecting?

@miketheman Do we need to change either of these queries to optimize them further? My SQLAlchemy-fu is a bit lacking so any suggestions would be helpful :pray:

File: warehouse/oidc/tasks.py
Comment (miketheman) [inline warehouse/oidc/tasks.py:23]: **thought:** ‏I'm having some difficulty in getting this query to run in a shell:

```shellsession
$ make shell
>>> from warehouse.packaging.models import File, Project, Release
>>> db.query(Project.id).distinct().join(Project.oidc_publishers)
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/code.py", line 90, in runcode
    exec(code, self.locals)
  File "<console>", line 1, in <module>
AttributeError: type object 'Project' has no attribute 'oidc_publishers'
```

I'm not entirely certain why this relationship isn't being expressed. 🤔 

Were you able to run this query?

Comment (miketheman) [reply warehouse/oidc/tasks.py:23]: Never mind, had to call `sqlalchemy.orm.configure_mappers()` after import to create the mappings, something that happens automatically in production.

File: warehouse/oidc/tasks.py
Comment (miketheman) [reply warehouse/oidc/tasks.py]: > Do we need to change either of these queries to optimize them further?

I'm gonna say "not yet" for now - let's let the task run and see how it does.

**Review by miketheman — APPROVED** [review]

**Comment by miketheman** [comment]
@sethmlarson Any objections about shipping this Monday instead of today, just in case the queries tax the db too heavily?

**Comment by sethmlarson** [comment]
@miketheman No objections here! We should maybe even wait for Wednesday since Monday and Tuesday are a holiday for many folks in the US so I and Ee won't be working that day.

File: warehouse/oidc/tasks.py
Comment (sethmlarson) [reply warehouse/oidc/tasks.py:23]: Glad you figured that out, I wouldn't have been much help there :P All I had was "unit tests seem to work!"

**Comment by sethmlarson** [comment]
@miketheman Looks like CI can't proceed without maintainer approval? Also I'm cool to deploy this now, I can create a revert PR in case we need to quickly revert the commit.

**Pull request merged**
